Provide a way to omit Event stages in audit policy

CaoShuFeng opened the pull request:
This provide a way to omit some stages for each audit policy rule.
    
For example:

    ```
      apiVersion: audit.k8s.io/v1beta1
      kind: Policy
      - level: Metadata
        resources:
           - group: "rbac.authorization.k8s.io"
             resources: ["roles"]
        omitStages:
          - "RequestReceived"
    ```
    
RequestReceived stage will not be emitted to audit backends with previous config.


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
None
```
#

CaoShuFeng commented:
@sttts @tallclair 
I will add unit test for this change soon if my idea is "right".

tallclair commented:
@CaoShuFeng - What is your motivation for omitting stages? Would the solution proposed in https://github.com/kubernetes/kubernetes/pull/46592#issuecomment-316553044 work for you? I'd like to explore that option before expanding the audit profile API.

CaoShuFeng commented:
@tallclair Thanks for your review.

I write this pr because of this [expression](https://github.com/kubernetes/kubernetes/issues/48561#issue-241059290):
> We want to omit the RequestReceived stage in GKE. The policy may be the right place to specify that.

It looks like I got it wrong.
What's the meaning behind `The policy may be the right place to specify that.` ?

tallclair commented:
This isn't wrong, per say. I just wanted to discuss alternatives before putting the burden on the user (i.e. adding it to the policy). If we can accomplish the same goals automatically (the batching proposal) then I think that would be preferable.

CaoShuFeng commented:
> . I just wanted to discuss alternatives before putting the burden on the user (i.e. adding it to the policy)

Agree.

Thanks for teaching me.
I will leave this pr open for a while. `Maybe` this is usefule.

caesarxuchao commented:
/assign @tallclair 
/unassign @caesarxuchao

CaoShuFeng commented:
Closed according to commants from @tallclair and @ericchiang .
Ref:
https://github.com/kubernetes/kubernetes/issues/48561#issuecomment-318146110
https://github.com/kubernetes/kubernetes/issues/48561#issuecomment-318163694

CaoShuFeng commented:
Repopened according to @crassirostris 's commants.
Ref:
https://github.com/kubernetes/kubernetes/issues/48561#issuecomment-318432155

This is still WIP, Unit tests will come soon.

crassirostris commented:
@tallclair suggested earlier that it might be useful to have `omitStages` field in the `Policy` object that will be "ored" with the `omitStages` field in the `PolicyRule`. That's a syntax sugar for cases where a stage has to be excluded in every rule

@sttts @CaoShuFeng WDYT?

crassirostris commented:
Btw, it should probably should be rebased once https://github.com/kubernetes/kubernetes/issues/49115 is merged and the changes should go to the beta API

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go:
```diff
@@ -528,6 +528,21 @@ func (m *PolicyRule) MarshalTo(dAtA []byte) (int, error) {
 			i += copy(dAtA[i:], s)
 		}
 	}
+	if len(m.OmitStages) > 0 {
```
no need for the if.

sttts replied:
oops, that's generated proto :)

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go:
```diff
@@ -723,6 +738,12 @@ func (m *PolicyRule) Size() (n int) {
 			n += 1 + l + sovGenerated(uint64(l))
 		}
 	}
+	if len(m.OmitStages) > 0 {
```
no need for the if

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go:
```diff
@@ -40,21 +40,21 @@ func NewChecker(policy *audit.Policy) Checker {
 }
 
-func (p *policyChecker) Level(attrs authorizer.Attributes) audit.Level {
+func (p *policyChecker) Level(attrs authorizer.Attributes) (audit.Level, []audit.Stage) {
```
don't the level and stages overlap?

CaoShuFeng replied:
Renamed to LevelAndStages

sttts replied:
<nit-picking>LevelAndOmitStaged</nit-picking>

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -200,6 +200,11 @@ type PolicyRule struct {
+
+	// OmitStages specify events generated in which stages will not be emitted to backend
+	// Allowed stages are RequestReceived,ResponseStarted,ResponseComplete,Panic
+	// empty list means every events will be emitted.
```
nit: An empty....

CaoShuFeng replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -200,6 +200,11 @@ type PolicyRule struct {
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend
+	// Allowed stages are RequestReceived,ResponseStarted,ResponseComplete,Panic
```
nit: fullstop and spaces after ,

sttts replied:
Is it common to relist the constants in those comments?

crassirostris replied:
> Is it common to relist the constants in those comments?

+1, I don't think it's a great idea

CaoShuFeng replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -157,9 +163,10 @@ var _ http.ResponseWriter = &auditResponseWriter{}
+	event      *auditinternal.Event
+	once       sync.Once
+	sink       audit.Sink
+	omitStages []auditinternal.Stage
```
please add a comment: omitStages are skipped, i.e. not sent to the backend. But their information is still recorded in the context Event for later stages.

CaoShuFeng replied:
Done.

sttts commented:
Some nit. Overall lgtm.

sttts commented:
> @tallclair suggested earlier that it might be useful to have omitStages field in the Policy object that will be "ored" with the omitStages

Sounds good.

CaoShuFeng commented:
> @tallclair suggested earlier that it might be useful to have omitStages field in the Policy object that will be "ored" with the omitStages

+1 will add it.

CaoShuFeng commented:
Unit tests has beed added to it.


omitStages in Policy Object will be added later. And this needs a rebase after #49115

CaoShuFeng commented:
@crassirostris 
Can you help me start the tests for this pr?
This needs a "/ok-to-test" label.
So that I can see whether it works well.

crassirostris commented:
/ok-to-test

CaoShuFeng commented:
I will rebase this change after https://github.com/kubernetes/kubernetes/pull/48836 gets merged.

sttts commented:
@CaoShuFeng please rebase and run hack/update-gofmt.sh.

/lgtm
/approve no-issue

soltysh commented on staging/src/k8s.io/apiserver/pkg/audit/policy/checker_test.go:
```diff
@@ -111,52 +111,73 @@ func TestChecker(t *testing.T) {
+				audit.Stage("RequestReceived"),
+			},
+		},
+		"only audit painc": {
```
s/painc/panic

soltysh reviewed (commented):
Mind fixing the typo while rebasing this PR

CaoShuFeng commented:
I will rebase this after https://github.com/kubernetes/kubernetes/pull/51119 get merged.
@soltysh

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -92,65 +116,35 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
-// - context with audit event attached to it
-// - created audit event
-// - error if anything bad happened
-func createAuditEventAndAttachToContext(requestContextMapper request.RequestContextMapper, req *http.Request, policy policy.Checker) (request.Context, *auditinternal.Event, error) {
```
@soltysh sorry for tearing down this function. :(

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce

sttts commented:
Needs rebase.

CaoShuFeng commented:
> Needs rebase.

Partially done. This will need another rebase after #51797.
#51797 has gone into the submit queue.

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce-bazel

sttts commented:
/retest

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -42,19 +42,43 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 		return handler
 	}
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		ctx, ev, err := createAuditEventAndAttachToContext(requestContextMapper, req, policy)
```
Can't we return omitStages from `createAuditEventAndAttachToContext`? Am not a fan of this code duplication.

crassirostris replied:
I'm not sure why this function is removed at all. `omitStages` are later obtained from the `policy` and not used until the code that was previously outside of `createAuditEventAndAttachToContext`. @CaoShuFeng Could you please elaborate more?

CaoShuFeng replied:
I have took `createAuditEventAndAttachToContext` back.
And the codes look cleaner for me, thanks.

sttts replied:
much better now

sttts replied:
lgtm. @crassirostris can you take another look and /lgtm if it is ok?

crassirostris commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -207,6 +207,10 @@ type PolicyRule struct {
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend.
+	// An empty list means every events will be emitted.
```
I think that might be confusing. What about `An empty list means no restrictions will apply.`?

CaoShuFeng replied:
Done.

crassirostris commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -173,6 +179,8 @@ type auditResponseWriter struct {
 	event *auditinternal.Event
 	once  sync.Once
 	sink  audit.Sink
+	// omitStages are skipped, i.e. not sent to the backend. But their information is still recorded in the context Event for later stages.
```
> omitStages are skipped

Is this comment necessary? `once` and `sink` are not set to the backend also :) Basically, the only explanation why `omitStages` field is there is `b/c it's used later` and this description is not very informative. Maybe remove the comment altogether?

CaoShuFeng replied:
Good for me.
Deleted.

crassirostris reviewed (commented):
Some nits regarding the comments, otherwise LGTM

CaoShuFeng committed:
Provide a way to omit Event stages in audit policy

Updates https://github.com/kubernetes/kubernetes/issues/48561
This provide a way to omit some stages for each audit policy rule.

For example:
  apiVersion: audit.k8s.io/v1beta1
  kind: Policy
  - level: Metadata
    resources:
       - group: "rbac.authorization.k8s.io"
         resources: ["roles"]
    omitStages:
      - "RequestReceived"

RequestReceived stage will not be emitted to audit backends with
previous config.

crassirostris reviewed (approved):
LGTM

CaoShuFeng committed:
generated: update API resources

./hack/update-codegen.sh
./hack/update-generated-protobuf.sh

crassirostris commented:
/lgtm

tengqm commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -207,6 +207,10 @@ type PolicyRule struct {
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend.
+	// An empty list means no restrictions will apply.
```
should we mark this as optional?

CaoShuFeng replied:
Yes, we should.

CaoShuFeng replied:
Added here together:
https://github.com/kubernetes/kubernetes/pull/54634
